env0/resource_environment.go

```diff
@@ -0,0 +1,184 @@
+				Computed:    true,
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Description: "The environment's name",
+				Required:    true,
+			},
+			"project_id": {
+				Type:        schema.TypeString,
+				Description: "the environment's project id",
+				Required:    true,
+			},
+			"template_id": {
+				Type:        schema.TypeString,
+				Description: "the environment's template id",
+				Required:    true,
+			},
+		},
+	}
+}
+
+func setEnvironmentSchema(d *schema.ResourceData, environment client.Environment) {
+	d.Set("name", environment.Name)
+	d.Set("project_id", environment.ProjectId)
+	d.Set("template_id", environment.TemplateId)
+}
+
+func resourceEnvironmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	payload := getCreatePayload(d)
+
+	environment, err := apiClient.EnvironmentCreate(payload)
+	if err != nil {
+		return diag.Errorf("could not create environment: %v", err)
+	}
+
+	d.SetId(environment.Id)
+	setEnvironmentSchema(d, environment)
+
+	return nil
+}
+
+func resourceEnvironmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	environment, err := apiClient.Environment(d.Id())
+	if err != nil {
+		return diag.Errorf("could not get environment: %v", err)
+	}
+
+	setEnvironmentSchema(d, environment)
+
+	return nil
+}
+
+func resourceEnvironmentUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	payload := getUpdatePayload(d)
+
+	// TODO: deploy if needed
+
+	_, err := apiClient.EnvironmentUpdate(d.Id(), payload)
+	if err != nil {
+		return diag.Errorf("could not update environment: %v", err)
+	}
+
+	return nil
+}
+
+func resourceEnvironmentDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	_, err := apiClient.EnvironmentDestroy(d.Id())
+	if err != nil {
+		return diag.Errorf("could not delete team: %v", err)
+	}
+	return nil
+}
+
+func getCreatePayload(d *schema.ResourceData) client.EnvironmentCreate {
+	payload := client.EnvironmentCreate{}
+
+	if name, ok := d.GetOk("name"); ok {
+		payload.Name = name.(string)
+	}
+
+	if projectId, ok := d.GetOk("project_id"); ok {
+		payload.ProjectId = projectId.(string)
+	}
+
+	payload.DeployRequest = &client.DeployRequest{}
+
+	if blueprintId, ok := d.GetOk("template_id"); ok {
+		payload.DeployRequest.BlueprintId = blueprintId.(string)
+	}
+
+	if blueprintRepository, ok := d.GetOk("repository"); ok {
+		payload.DeployRequest.BlueprintRepository = blueprintRepository.(string)
+	}
+
+	if blueprintRevision, ok := d.GetOk("revision"); ok {
+		payload.DeployRequest.BlueprintRepository = blueprintRevision.(string)
+	}
+
+	return payload
+}
+
+func getUpdatePayload(d *schema.ResourceData) client.EnvironmentUpdate {
```

Comment:
Why create and update are treated differently when creating payload? 🤔

Reply:
when it comes to environments different fields updates should invoke different api endpoints (trigger redeployment \ update ttl for example)

you can see in [our api](https://developer.env0.com/docs/api/b3A6MTkzNDg1MTk-create-environment) that most of whats on create environment isn't on update environment, this will come in to play in [part 2](https://trello.com/c/K68z3UXy/4784-terraform-provider-environment-resource-part-2)